60 seed data

ski-climb opened the pull request:
@tmikeschu @njgheorghita @Sh1pley @riverswb Don't merge this one yet, but definitely take a look.

Worth noting: this branch will require the branch that @riverswb has (removing the reservations foreign key on the nights model) be merged in before this one.

Soooo, after I pulled the trigger and merged the branch for making reservations, I realized that, although the tests passed, none of it worked in my browser.  Many sleuthings later... it seems that the date-picker that we're using provides dates in the mm/dd/yyyy format, but we set up some of our tests before we had the date-picker in place, so we were dealing with dates in the dd-mm-yyyy format.  Very understandable on both counts, but now we have to make it work.

Bottom line is the `Date.parse(some-date)` doesn't play well with mm/dd/yyyy format that we get from the date-picker.

The commit `792de3c` makes the change in our code to the mm/dd/yyyy format for passing around dates (so that the tests will more accurately represent what's happening in the browser).  This broke many of the specs, so those were rewritten too.

It came down to a choice of either letting the `Night` model accept only a very restricted date format, or assume that the `Night` model would always be provided with check_in and check_out dates which were `Date` objects or played well with `Date.parse(some_date)`.  I chose the latter, but please let me know if you think the trade-off makes more sense the other way.  I'm open minded.

ski-climb committed:
Use progress bar when seeding data

ski-climb committed:
Use longer names for couch factory to avoid confusion

ski-climb committed:
Use consistent password for user factory for developer happiness

ski-climb committed:
Allow Night check in/out scope to take date objects too

ski-climb committed:
WIP Seed data for users, couches, and nights

ski-climb committed:
Merge branch 'master' into 60-seed-data

ski-climb committed:
Update specs to work with datepicker date format

ski-climb committed:
Make seed file object oriented

ski-climb committed:
Scope nights to those which are available

ski-climb committed:
Correct couches factory for valid street addresses

ski-climb committed:
Travelers have reservations in seed file

tmikeschu commented on Gemfile:
```diff
@@ -23,6 +23,7 @@ gem 'uglifier', '>= 1.3.0'
 gem 'coffee-rails', '~> 4.2'
 # See https://github.com/rails/execjs#readme for more supported runtimes
 # gem 'therubyracer', platforms: :ruby
+gem 'progressbar'
```
@ski-climb shiny buttons! What does it do?

ski-climb replied:
When you run `rails db:seed` it will show a progress bar across your terminal as it creates new objects in the database.  The whole process takes... maybe 20-30 seconds?

tmikeschu commented on app/models/couch.rb:
```diff
@@ -10,8 +10,14 @@ def self.search(params)
+  private
+
+    def self.to_date(string)
+      Date.strptime(string, "%m/%d/%Y")
```
Totally cool with this solution 👍

njgheorghita replied:
👍

tmikeschu commented on app/models/night.rb:
```diff
@@ -1,10 +1,11 @@
   scope :all_for_couch, -> (couch) { where(couch: couch) }
   scope :between_check_in_check_out, -> (check_in, check_out) {
-    where(date: (Date.parse(check_in))..(Date.parse(check_out) - 1.day))
+    where(date: (Date.parse(check_in.to_s))..(Date.parse(check_out.to_s) - 1.day))
```
@ski-climb what happens without this coercion?

ski-climb replied:
This might be a bit unnecessary.  Basically it allows us to pass strings (e.g. "2017-02-22") OR Date objects (e.g. `Date.current`) to the scope.  I was thinking about programming for least surprise, and it just felt right that this scope should be able to handle Date objects.

tmikeschu commented on db/seeds.rb:
```diff
@@ -1,7 +1,143 @@
+  end
+
+  def remove_all_existing_data
+    puts "Clearing out the data currently in your database (Ctrl-C now if that's a bad thing...)"
```
Clever!

tmikeschu commented on db/seeds.rb:
```diff
@@ -1,7 +1,143 @@
+    end
+
+    # generate user for developer log-in
+    @sample_user = FactoryGirl.create(:user, email: "user@example.com", phone_number: 111_111_1111)
```
👍 thinking ahead

tmikeschu commented on db/seeds.rb:
```diff
@@ -1,7 +1,143 @@
+  def create_users
+    number_of_users = 70
+
+    user_bar = ProgressBar.create(title: "Users", total: number_of_users)
```
Can't wait to see this progress bar in action!

tmikeschu commented on db/seeds.rb:
```diff
@@ -1,7 +1,143 @@
+  attr_reader :sample_user
+
+  def self.start
+    seed = Seed.new
```
👌

tmikeschu commented on db/seeds.rb:
```diff
@@ -1,7 +1,143 @@
+    CSV.foreach(filename, headers: true) do |row|
+
+      couch_attributes = {
+        name: Faker::Ancient.god + " " + Faker::Superhero.suffix + " of " + Faker::HarryPotter.location,
```
@ski-climb you had too much fun 😄

tmikeschu commented on db/seeds.rb:
```diff
@@ -1,7 +1,143 @@
+        night_bar.increment
+      end
+
+      number_of_contiguous_nights.sample.times do |i|
```
@ski-climb can we do an in person walk through on this tomorrow? Looks solid, just want to get a more in depth explanation.

njgheorghita replied:
ditto, would also find that helpful

ski-climb replied:
@tmikeschu @njgheorghita happy to!

tmikeschu commented on spec/features/reservations/traveler_can_make_a_reservation_spec.rb:
```diff
@@ -28,19 +29,22 @@
-      fill_in "Check In", with: today
-      fill_in "Check Out", with: tomorrow
+
+      fill_in "Check In", with: today.to_date_picker_format
```
👍

tmikeschu commented on spec/models/reservation_maker_spec.rb:
```diff
@@ -7,7 +8,11 @@
     couch.nights << create(:night, date: Date.tomorrow.tomorrow)
     @traveler = create(:user)
-    @nights_params = { "check_in" => Date.current.to_s, "check_out" => Date.tomorrow.tomorrow.to_s, "couch_id" => couch.id }
+    @nights_params = { 
```
Thanks for improving the readability as well!

tmikeschu reviewed (commented):
Super thorough fix on the date problem, and nice work with the seed!

Ready to merge once @riverswb 's PR is in.

njgheorghita commented on db/seeds.rb:
```diff
@@ -1,7 +1,143 @@
+  def remove_all_existing_data
+    puts "Clearing out the data currently in your database (Ctrl-C now if that's a bad thing...)"
+    sleep(5)
+    puts "Too late!"
```
Lolz

njgheorghita commented:
Killer @ski-climb, thanks for taking care of this, definitely gonna be a massive help with dev and production. Not sure where we are on the merge status, still waiting for the night model pr?? either way it's good to merge asap

tmikeschu commented on spec/features/reservations/traveler_can_make_a_reservation_spec.rb:
```diff
@@ -10,6 +10,7 @@
 =end
 
 require "rails_helper"
+include ModelHelpers
```
@ski-climb I think this can be deleted because of your rails helper setup.

ski-climb replied:
I thought so too... but then I was getting errors for not having the `.to_date_picker_format` method defined.  Perhaps I have it setup wrong.  Another set of eyes would be much appreciated!

ski-climb committed:
Merge master with 60-ne-seed-data

ski-climb committed:
Git merge master into 60-ne-seed-data

I ended up doing this twice.  I was trying to ensure that the merge went
smoothly be first merging (locally) my branch into master.

It was not the smooth process I had hoped for, and I basically ended up
resolving the (slightly different) merge conflicts twice.  Not the most
fun I've had on a Sunday night.

ski-climb committed:
Update for Nights validation on not creating dates in the past

ski-climb committed:
Merge branch 'master' into 60-seed-data

ski-climb committed:
Update seeds file to make nights in the past - despite validation

ski-climb committed:
Rewrite how Nights are updated/saved in seeds file for validation

tmikeschu reviewed (approved):
Nice work, @ski-climb ! See comment about needing to include your model helpers module. 


Thumbs up for the next reviewer to merge.

riverswb reviewed (approved):
Great work! Merging
